### ISSUE TITLE
nodeSets and Elasticsearch autoscaling policy API

### ISSUE THREAD

**Issue by barkbay** [elastic/cloud-on-k8s#3955]
The purpose of this issue is to discuss how the new autoscaling policy API integrates with the concept of `nodeSet` _(the full content of the autoscaling spec and where it should live, as a new CRD or as an annotation, can be discussed in another issue)_.

We already discussed it a bit internaly and while several ideas have emerged we have identified some important driving factors:

* We want to be able to iterate on the Elasticsearch resource definition (a.k.a. CRD) without the burden of introducing a new version. A consequence is that it's not possible to tweak the current `Elasticsearch` resource definition, for example it prevents to add new fields in the `nodeSet` definition to create references to the autoscaling policies.

* In the context of the Elasticsearch autoscaling API the set of roles applied to an Elasticsearch node is important and meaningful. It is an argument in favor of having those roles explictly set in the `nodeSet` configuration.

It leaves us with 2 ways to specify an association between a `nodeSet` and an autoscaling policy:

1. The set of roles is duplicated in the node configuration and in the autoscaling policy definition. That first option can be seen as a kind of "pattern matching" between the roles in the node configuration and the roles in the autoscaling policy specification. An autoscaling policy would be applied to any `nodeSet` which has the exact same set of roles.
2. A dedicated field in the autoscaling policy specification is set to explicitly create a symbolic reference to a group of `nodeSets`.

To make it more concrete here are two examples _(`autoscalingPolicies` specification appears in the Elasticsearch CRD only to make it easier to read, it does not preclude if it will eventually be in a dedicated CRD or in a annotation) :_

### Option 1: Implicit association

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
 name: quickstart
spec:
 version: 7.10.0
 nodeSets:
   # Declare a dedicated nodeSet to deploy some nodes in europe-west3-a
   - name: data-west3-a
     config:
       roles:  [ "master", "data_hot", "ingest" ]
        # Declare a dedicated nodeSet to deploy some nodes in europe-west3-b
   - name: data-west3-b
     config:
       roles:  [ "master", "data_hot", "ingest" ]
 # Autoscaling policies are defined here:
 autoscalingPolicies:
 - name: master_data_ingest
   # Policy only applies to nodeSets with the exact same set of roles
   roles:  [ "master", "data_hot", "ingest" ]
   minAllowed:
      count: 2
      cpu: 16
      memory: 8Gi
      storage: 10G
   maxAllowed:
      count: 6
      cpu: 16
      memory: 8Gi
      storage: 100G
...
```

### Option 2: Explicit association using a nodeSet reference

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
 name: quickstart
spec:
 version: 7.10.0
 nodeSets:
   # Declare a dedicated nodeSet to deploy some nodes in europe-west3-a
   - name: data-west3-a
     config:
       roles:  [ "master", "data_hot", "ingest" ]
        # Declare a dedicated nodeSet to deploy some nodes in europe-west3-b
   - name: data-west3-b
     config:
       roles:  [ "master", "data_hot", "ingest" ]
 # Autoscaling policies are defined here:
 autoscalingPolicies:
 - name: master_data_ingest
   # these are the nodeSets to which the autoscaling policy applies
   nodeSetRefs : ["data-west3-a" , "data-west3-b"]
   minAllowed:
      count: 2
      cpu: 16
      memory: 8Gi
      storage: 10G
   maxAllowed:
      count: 6
      cpu: 16
      memory: 8Gi
      storage: 100G
...
```

The first option seems pretty straightforward and consistent with the Elasticsearch autoscaling API (the set of roles is duplicated in both the node configuration and in the autoscaling policy). A drawback is that the user must keep all the set of roles in sync between the various `nodeSet` definitions and the autoscaling policy specification.

The second option avoids the user to duplicate the set of roles, but it still has to remember to add any new `nodeSets` in the autoscaling policy specification. It also requires another level of validation from ECK to ensure that the set of roles are consistent across the referenced `nodeSets`.

Overall I think I would be in favor of starting with the first option, but glad to hear if anyone has another opinion/comment.

CC @bleskes

**Comment by sebgl** [comment]
tl;dr: I'm in favor of option 1.

Both options seem to have the same drawbacks: the user needs to make sure what they set in the `autoscalingPolicies` section matches what they set in the `nodeSet` section. They may run into troubles if one is not consistent with the other.

I like the first option better because I think it makes things more explicit (`master, data_hot, ingest` nodes will be autoscaled, regardless of how nodeSets are setup).
The 2nd option leaves more place to confusion and unintended behaviour IMO. For example if the user specifies:

```
 nodeSets:
   - name: master-nodes
     config:
       roles:  [ "master" ]
   - name: data-west-a
     config:
       roles:  [ "data_hot", "ingest" ]
   - name: data-west-b
     config:
       roles:  [ "data_hot", "ingest" ]
 autoscalingPolicies:
 - name: default
   nodeSetRefs : ["master-nodes" , "data-west-a", "data-west-b"]
   minAllowed: {}
   maxAllowed: {}
```

When I read this spec (from a user perspective) then I don't know what's going to happen.
Will each nodeSet be scaled independently (-> 3 policies behind the scenes), or will they be implicitly grouped based on their roles (-> 1 policy for masters, 1 policy for data noes), or will that lead to a single policy for all nodes (master + data treated as the same group)? I don't know, so I'll need to spend quite some time reading the documentation to understand how that API works.
We can of course have ECK reject that specification, with an error that implies the spec is invalid because the implicit node roles (through nodeSet refs) in my policy does not match the roles of my nodeSets. At that point I have to change my mental model to express things in terms of node roles in my head, then translate that into nodeSets into the spec (implicitly group nodeSets by node roles) to make sure what I write is correct.

The first option removes those doubts. It's very clear that the autoscaling will be based on the node roles:

 ```
 nodeSets:
   - name: master-nodes
     config:
       roles:  [ "master" ]
   - name: data-west-a
     config:
       roles:  [ "data_hot", "ingest" ]
   - name: data-west-b
     config:
       roles:  [ "data_hot", "ingest" ]
 autoscalingPolicies:
 - name: master-policy
   roles : ["master"]
   minAllowed: {}
   maxAllowed: {}
 - name: data-policy
   roles : ["data_hot"]
   minAllowed: {}
   maxAllowed: {}
```

The main drawback I see in this approach in the above example: the `data-policy` policy specified here is useless (it's missing the `ingest` role to match my data nodes!), which is something we can warn the user about (through a k8s event?) very explicitly: `the data-policy does not match any node`. When the user reads that error it seems quite easy to understand how to fix the spec.

You called option 1 "implicit" and option 2 "explicit". It's funny because I see things the other way around (specifying node roles is explicit, specifying nodeSets is implicit).

**Comment by pebrc** [comment]
I think I agree with @sebgl here. Looking at this again the matching on roles feels more natural. As I mentioned out of band before I think we should stay close the Elasticsearch API to avoid unnecessary cognitive overhead. 

If we agree on this approach then usability is the main challenge, we can generate events and log statements which all have good chances of being missed by users. Since k8s 1.19 admission Webhooks can also return warnings which might be an interesting feature for this kind of problem: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admissionreview-response-warning But of course that only covers users on the latest version of k8s.

**Comment by barkbay** [comment]
I'm closing this issue as it seems that we have an agreement on option 1:

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
 name: quickstart
spec:
 version: 7.10.0
 nodeSets:
   # Declare a dedicated nodeSet to deploy some nodes in europe-west3-a
   - name: data-west3-a
     config:
       roles:  [ "master", "data_hot", "ingest" ]
        # Declare a dedicated nodeSet to deploy some nodes in europe-west3-b
   - name: data-west3-b
     config:
       roles:  [ "master", "data_hot", "ingest" ]
 # Autoscaling policies are defined here:
 autoscalingPolicies:
 - name: master_data_ingest
   # Policy only applies to nodeSets with the exact same set of roles
   roles:  [ "master", "data_hot", "ingest" ]
   minAllowed:
      count: 2
      cpu: 16
      memory: 8Gi
      storage: 10G
   maxAllowed:
      count: 6
      cpu: 16
      memory: 8Gi
      storage: 100G
...
```

**Comment by zanbel** [comment]
I know I'm very late to this party, but wanted to offer my opinion (although I don't feel strongly about it :))

I do think that for a "static" configuration option 1 works nicely since it's more explicit. But when users will update their configuration, for example, when adding dedicated master or ingest nodes, a new dedicated warm data tier, and so on, then they are more likely to forget to update their policy, and things will start to break. In that case, I think the more implicit behavior can come in handy.  In ECE/ESS, users don't really need to think about it. They define the relevant limits per tier/slider and are not concerned with these updates when adding new data tiers or new dedicated node types.

The option I thought about was for the policy to reference the nodeSets. That way, when I change the roles of the nodeSets, existing autoscaling behavior will not be affected. This feels more natural than the other way around since the more common flow will be to start with a single nodeSet and then adding data tiers and node types. For example, I can start with the following:
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
 name: quickstart
spec:
 version: 7.10.0
 nodeSets:
   - name: data-default
     config:
       nodeSets:  [ "master", "data_hot", "data-warm", "ingest" ]
 autoscalingPolicies:
 - name: data
   nodeSets:  [ "data-default" ]
   minAllowed:
      count: 2
      cpu: 16
      memory: 8Gi
      storage: 10G
   maxAllowed:
      count: 6
      cpu: 16
      memory: 8Gi
      storage: 100G
...
```

And then I want to add dedicated master nodes
```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
 name: quickstart
spec:
 version: 7.10.0
 nodeSets:
   - name: data-default
     config:
       nodeSets:  [ "data_hot", "data-warm", "ingest" ]
   - name: master
     config:
       nodeSets:  [ "master" ]
       # Some fixed master nodes configuration
       ...
 autoscalingPolicies:
 - name: data
   nodeSets:  [ "data-default" ]
   minAllowed:
      count: 2
      cpu: 16
      memory: 8Gi
      storage: 10G
   maxAllowed:
      count: 6
      cpu: 16
      memory: 8Gi
      storage: 100G
...
```

So then we update the policy roles on behalf of the user and they don't need to worry about it.

Having said that, I also mentioned to Michael, this actually might also be relevant feedback for @henningandersen. Per @pebrc 's comment about aligning with ES API, which does exactly match between the roles defined in the policy and node's roles, is important. 

As mentioned above, I don't mean to re-open this for discussion but wanted to share my thoughts.
